proposals/0000-language-editions.rst

```diff
@@ -0,0 +1,1144 @@
+design, readers may want to skim the introductory sections of that manifesto.
+However, the effective payload of this proposal is all in this one document.
+
+Motivation
+----------
+
+The primary motivation behind the use of language editions is that they
+can succinctly inform GHC what kind of user it's faced with, so GHC
+can behave accordingly.
+
+Though the details are spelled out below, it's necessary to introduce
+some of the language editions I'm proposing:
+
+* ``Stable2024``: Code compiled in the ``Stable2024`` edition will be
+  expected to compile (assuming stability of libraries) under that
+  same language edition on all GHCs release for 6 years,
+  until the beginning of 2030. (We may end up falling short of this
+  expectation in corner cases, but it will remain a reasonable expectation
+  to have.)
+
+* ``Experimental2024``: Switching to the ``Experimental`` series of
+  editions gives you access to experimental features of GHC, which might
+  reasonably evolve and break existing programs.
+
+* ``Latest``: This language edition is the latest and greatest that GHC
+  has to offer. Compiling with ``Latest`` might break between releases.
+
+* ``Student2024``: The student edition of the language has extra guardrails.
+  Furthermore, any code written in the ``Student`` edition is expected
+  to compile with all future versions of GHC for 10 years, meant to
+  echo the expected lifetime of a textbook.
+
+With that out of the way, here are some scenarios that the language editions
+model helps us to address:
+
+* **Stability**. Once we include a feature in a ``Stable`` edition, it
+  will not change. However, that does not mean GHC cannot evolve. If we
+  identify a ``Stable`` feature that nevertheless deserves an upgrade, we
+  can do so, provided we continue to support the old behavior. So
+  type inference around ``FlexibleInstances``, say, might work one way in
+  ``Stable2024`` and differently in ``Stable2027``. If we release a new
+  ``Stable`` edition every 3 years, users have three years to upgrade before
+  we no longer guarantee support. A three-year timeline is long enough that
+  we might imagine contributors writing upgrade tools in that timeframe.
+
+* **Error messages**. Because the language edition describes the user (very
+  coarsely!) we can tailor error messages to be appropriate for them. The
+  ``Student`` editions will not advise ``DataKinds``. The ``Stable`` editions
+  will not advise ``LinearTypes``.
+
+* **Warning evolution**. One challenge in adding features to GHC is to
+  know how to evolve the warning system. Does a new warning get added to
+  ``-Wall``? And what's the deal with ``-Wcompat``?
+
+  Language editions make
+  these questions easier to answer: If we think that a warning should be
+  enabled for users going forward, we can turn it on by default, but only
+  in appropriate language editions, such as ``Experimental`` or ``Latest``.
```

Comment:
Nothing is stopping us from adding warnings to existing stable language editions. I've added a sentence to the proposal to clarify this point. The key part of the text you've highlighted above is the "if": language editions give us the ability to add warnings to one edition but not another, if that's right for a particular warning. Anecdotally, my sense is that we've avoided putting warnings on by default because of worries about stability concerns; if we had had a way of saying "this is on by default in all new code", we would have. I don't have evidence to support this claim, sadly. And it might be a faulty understanding or memory.

Reply:
> Anecdotally, my sense is that we've avoided putting warnings on by default because of worries about stability concerns

Fortunately, we now have [stability principles](https://github.com/ghc-proposals/ghc-proposals/blob/master/principles.rst?plain=1#L397) that explicitly state that warning stability is not a goal. Let's follow them!